### PR TITLE
Daisy Chain PR

### PR DESCRIPTION
**Pull request by LorenLuke**
Enables Daisy Chaining of explosives together.

### CODE REVIEW

File: addons/explosives/stringtable.xml [hunk 1]
```diff
@@ -505,6 +505,7 @@
             <Italian>Raccogli</Italian>
             <Portuguese>Pegar</Portuguese>
         </Key>
+<<<<<<< HEAD
```
Comment (PabstMirror) [inline addons/explosives/stringtable.xml]: `<<<<< HEAD` - from merge conflict

for stringtables delete lines that aren't translated

**Comment by alganthe** [comment]
Example on how to use
- Place 3 explosive charges
- Set the first to detonate with a clacker
- Set the two other to detonate with daisy chain
- Start daisy chain on the one with the clacker detonation
- Attach daisy chain to explosive 2
- Start daisy chain on explosive 1
- Attach daisy chain to explosive 3

you can daisy chain any type of explosive, you need to start a daisy chain each time you want to link a new explosive.

**Comment by CorruptedHeart** [comment]
I will take a look at this, in around 8-10 hours time.

**Comment by LorenLuke** [comment]
If the only unintended difference is the stringtables, go ahead and simply use the new ones, but don't forget the daisy chain strings.

**Comment by CorruptedHeart** [comment]
Unfortunately didn't get a chance to look at this yet. Due to the joy that is loadshedding (love South Africa). Hopefully around 8 GMT+2 I will get some time to look at this more indepth.

File: addons/explosives/XEH_preInit.sqf [hunk 2]
```diff
@@ -60,4 +60,13 @@ PREP(startDefuse);
 PREP(startTimer);
 PREP(triggerType);
 
+PREP(canDaisyChainTo);
```
Comment (CorruptedHeart) [inline addons/explosives/XEH_preInit.sqf]: These will all need to be done alphabetically like the rest of the file please.

File: addons/explosives/functions/fnc_canDaisyChainTo.sqf [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * Return Value:
+ * Able to Daisy Chain to<BOOL>
+ *
+ * Example:
+ * if ([player,_explosive] call ACE_Explosives_fnc_canDaisyChain) then {hint "Can Daisy Chain";};
+ *
+ * Public: Yes
+ */
+#include "script_component.hpp"
+private ["_explosive","_target","_explosive","_result"];
```
Comment (CorruptedHeart) [inline addons/explosives/functions/fnc_canDaisyChainTo.sqf]: `_explosive` is defined twice and `_unit` isn't defined at all.

File: addons/explosives/functions/fnc_canDaisyChainTo.sqf [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+_unit = _this select 0;
```
Comment (CorruptedHeart) [inline addons/explosives/functions/fnc_canDaisyChainTo.sqf]: Can use `PARAMS(_unit,_target)` here and then you won't need to define `_unit` or `_target` above.

File: addons/explosives/functions/fnc_checkDaisyChain.sqf [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "script_component.hpp";
+
+private ["_explosive","_target","_range","_result"];
+_explosive = _this select 0;
+_target = _this select 1;
+_range = _this select 2;
+_result = false;
+
+if (_explosive distance _target < _range) then {_result = true};
```
Comment (CorruptedHeart) [inline addons/explosives/functions/fnc_checkDaisyChain.sqf]: Can use an `exitWith{true};` here and a false at then end. Might also want to wrap the distance part in `()`

File: addons/explosives/functions/fnc_daisyChainTo.sqf [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    GVAR(DaisyChains) = [];
+};
+
+_result = false;
+
+{
+    if (count GVAR(DaisyChains) < 1) exitWith {};
+    if (_x select 0 == _explosive) then {
+	    if (_x select 1 == _target) then {_result = true};
+	};
```
Comment (CorruptedHeart) [inline addons/explosives/functions/fnc_daisyChainTo.sqf]: I think there's some weird tabbing going on here.

File: addons/explosives/functions/fnc_detonateDaisyChain.sqf [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+#include "script_component.hpp"
+   
+private ["_explosive","_time"];
+
+_explosive = _this select 0;
+_time = _this select 1;
+
+ 
+if (isNil QGVAR(DaisyChains)) then {
```
Comment (CorruptedHeart) [inline addons/explosives/functions/fnc_detonateDaisyChain.sqf]: Can probably simply exit here.

File: addons/explosives/functions/fnc_detonateDaisyChain.sqf [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    GVAR(DaisyChains) = [];
+};
+
+{
+    if (_x select 0 == _explosive) then {
+    	_temparr = [_explosive, _x select 1];
+	    GVAR(DaisyChains) = GVAR(DaisyChains) - [_x];
+		if ([_explosive, _temparr select 1, 25] call FUNC(checkDaisyChain)) then {
+		    [_explosive, -1, [_temparr select 1, _time]] call FUNC(detonateExplosive);
+        };
```
Comment (CorruptedHeart) [inline addons/explosives/functions/fnc_detonateDaisyChain.sqf]: Some crazy indentation happening here.

File: addons/explosives/functions/fnc_isDaisyChaining.sqf [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * The unit is daisy chaining <BOOL>
+ *
+ * Example:
+ * _isdaisychaining = [player] call ACE_Explosives_fnc_isDaisyChaining;
+ *
+ * Public: Yes
+ */
+#include "script_component.hpp"
+
+private ["_unit", "_result", "_magazines"];
```
Comment (CorruptedHeart) [inline addons/explosives/functions/fnc_isDaisyChaining.sqf]: Unnecessary privates. `_magazines`

**Comment by CorruptedHeart** [comment]
So I finally had some time to take a look at the code. And while it appears that it does solve the daisy chaining issue. I'm not entirely sure the implementation is the best way to go about it.

Besides the issues I commented on, there's a couple others. A single global variable for daisy chaining might work. But that means that it is not possible to have multiple daisy chained explosives. So while this does at the base level solve the issue. There might be a better way. To discover that way, well that needs some thought.

**Comment by LorenLuke** [comment]
>  A single global variable for daisy chaining might work.

All daisy chains are stored via a global array in a commutative manner.
So when you put a daisy chain between explosives numbered for convenience:
1 daisy chained to 2, GVAR(DaisyChains) looks like 
[[1, 2], [2, 1]]
Include the above and add 2 chained to 3, it'll look like this:
[[1,2], [2, 1], [2, 3], [3, 2]] 
^so if 2 were to be set off, it would detonate 1 and 3.

Additionally, the instructions do daisy chain three of them together are:
interact with 1, start Daisy Chain, move to 2, attach Daisy Chain;
interact with 2, start Daisy Chain, move to 3, attach Daisy Chain.

So you can string together as many as you like, so long as they're in range.

File: addons/explosives/functions/fnc_checkDaisyChain.sqf (hunk 5 above)
Comment (ulteq) [inline addons/explosives/functions/fnc_checkDaisyChain.sqf]: You can do this in two lines:

`PARAMS_3(_explosive,_target,_range);`

`((_explosive distance _target) < _range)`

File: addons/explosives/functions/fnc_daisyChainTo.sqf [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+_target = _this select 1;
+
+if (isNil QGVAR(DaisyChains)) then {
+    GVAR(DaisyChains) = [];
+};
+
+_result = false;
+
+{
+    if (count GVAR(DaisyChains) < 1) exitWith {};
```
Comment (ulteq) [inline addons/explosives/functions/fnc_daisyChainTo.sqf]: I think this should be: `if (count _x < 1) exitWith {};`

File: addons/explosives/functions/fnc_daisyChainTo.sqf [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+_explosive = _this select 0;
+_target = _this select 1;
+
+if (isNil QGVAR(DaisyChains)) then {
+    GVAR(DaisyChains) = [];
+};
+
+_result = false;
+
+{
```
Comment (ulteq) [inline addons/explosives/functions/fnc_daisyChainTo.sqf]: You could just do this in the forEach body:

`if (_x isEqualTo [_explosive, _target]) exitWith { _result = true; };`

File: addons/explosives/functions/fnc_daisyChainTo.sqf [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    if (count GVAR(DaisyChains) < 1) exitWith {};
+    if (_x select 0 == _explosive) then {
+	    if (_x select 1 == _target) then {_result = true};
+	};
+} forEach GVAR(DaisyChains);
```
Comment (ulteq) [inline addons/explosives/functions/fnc_daisyChainTo.sqf]: You could just do this in the forEach body:

`if (_x isEqualTo [_explosive, _target]) exitWith { _result = true; };`

File: addons/explosives/stringtable.xml [hunk 13]
```diff
@@ -529,5 +530,103 @@
             <English></English>
             <Polish>Moduł ten pozwala dostosować opcje związane z ładunkami wybuchowymi, ich podkładaniem oraz rozbrajaniem.</Polish>
         </Key>
+=======
+        <Key ID="STR_ACE_Explosives_DaisyChainName">
+            <English>Daisy Chain</English>
+            <German>Daisy Chain</German>
```
Comment (esteldunedain) [inline addons/explosives/stringtable.xml]: @LorenLuke, just leave the english entry here and remove the rest. That way it's easier for traslators to find the missing entries to work on.

**Comment by ulteq** [comment]
> A single global variable for daisy chaining might work. But that means that it is not possible to have multiple daisy chained explosives.

Not entirely true.

`GVAR(DaisyChains) = [];`
`_daisyChainOne = [[1,2],[2,3],[3,4]];`
`_daisyChainTwo = [[5,6],[6,7]];`
`GVAR(DaisyChains) append _daisyChainOne;`
`GVAR(DaisyChains) append _daisyChainTwo;`

-> `GVAR(DaisyChains) = [[1,2],[2,3],[3,4],[5,6],[6,7]];`

If you trigger explosive 1, 2, 3 or 4 daisy chain one goes off.
If you trigger explosive 5, 6 or 7 daisy chain two goes off.

**Comment by ViperMaul** [comment]
Status?

**Comment by Glowbal** [comment]
Closing due to no activity and merge conflicts.

**Pull request closed without merging**